### PR TITLE
Fpga operators

### PR DESCRIPTION
**Pull request by ratusz**
FPGA first SGEMM operator

### CODE REVIEW

File: src/runtime/local/kernels/FPGAOPENCL/gemm_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+            serialized_A[addr++] = A[k + i*TOTAL_K];
+        }
+    // Serialize B
+    addr = 0;
+    for (int j = 0; j < TOTAL_J; j++)
+        for (int k = 0; k < TOTAL_K; k++) {
+            serialized_B[addr++] = B[j+k*TOTAL_J];
+        }
+#ifndef NDEBUG
+    DPRINTF("\n===== Host-CPU setting up the OpenCL platform and device ======\n\n");
```
Comment (corepointer) [inline src/runtime/local/kernels/FPGAOPENCL/gemm_interface.cpp:132]: All the setup code should be moved to a separate function that is called in the beginning of the Daphne execution, not on every sgemm()

**Review by corepointer — CHANGES_REQUESTED** [review]
Hi Piotr!
Thanks for your PR. The merge conflicts can be easily resolved if you just put your changes in front or behind the conflicting changes from main (which stem mainly from Arrow support doing the same thing as your FPGA support).

A few changes I'd like to suggest:
* remove unnecessary files:build.log, build2.sh, libFPGAOPENCLKernels.so
* move the example to the scripts/example subdir
* rename sourceme.sh to something more descriptive related to FPGA (e.g., fpga-build-env.sh ?)
* remove the absolute paths related to your setup from sourceme.sh. Maybe require the user to set paths in that file or set environment variables
* The path in FPGAOPENCL/CMakeLists.txt line 49 (target_link_libraries) is hard coded. Either find a way to read that from what CMake finds when you ask it to look for OpenCL support (that's probably the more difficult solution) or require to set that path by the user (similarly to the other paths that are set for the cmake command in build.sh)
* same for the include directory
* the gemm.aocx is a 42 MB binary file. Maybe we can have that as a downloadable dependency?
* More comments and suggestions in the code-relative comments above/below.

Best regards, Mark

File: src/runtime/local/kernels/FPGAOPENCL/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+# Specifies how to generate the file "kernels.cpp" (which resides in the build
+# directory) as the basis for the pre-compiled kernels library.
+
+#       * more properly set up paths (ugly relative include path below)
+#cmake_minimum_required(VERSION 3.21)
+
+#project(FPGAOPENCLkernels)
+
+include_directories(/opt/intel/intelFPGA_pro/21.4/hld/examples_aoc/common/inc/ /opt/intel/intelFPGA_pro/21.4/hld/host/include/)
```
[the remark this answers is not in the document]
Comment (ratusz) [reply src/runtime/local/kernels/FPGAOPENCL/CMakeLists.txt:23]: resolved by env variable

File: src/runtime/local/kernels/FPGAOPENCL/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	${PREFIX}/gemm_interface.h
+	${PROJECT_BINARY_DIR}/FPGAOPENCLkernels.cpp
+)
+
+add_library(FPGAOPENCLKernels SHARED ${FPGAOPENCLKernels_SRC})
+set_target_properties(FPGAOPENCLKernels PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/src/runtime/local/kernels)
+
+target_include_directories(FPGAOPENCLKernels PUBLIC ${PROJECT_SOURCE_DIR}/src/)
+
+target_link_libraries(FPGAOPENCLKernels PUBLIC AllKernels LLVMSupport /opt/intel/intelFPGA_pro/21.4/hld/linux64/lib/libOpenCL.so)
```
[the remark this answers is not in the document]
Comment (ratusz) [reply src/runtime/local/kernels/FPGAOPENCL/CMakeLists.txt:49]: resolved by env variable

File: sourceme.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+cd /home/mdokter/code/daphne-build-env/ ; source env.sh ; cd -
```
[the remark this answers is not in the document]
Comment (ratusz) [reply sourceme.sh:2]: removed

File: sourceme.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+#initialize Intel FPGA OpenCL environment
+source /opt/intel/intelFPGA_pro/21.4/hld/init_opencl.sh
+echo $INTELFPGAOCLSDKROOT
+export ALTERAOCLSDKROOT=$INTELFPGAOCLSDKROOT
+
+#set up BITSTREAM variable for required FPGA image (can be different different for varius implemented kernels)
+export BITSTREAM=/home/pratuszniak/pratuszn/daphne/src/runtime/local/kernels/FPGAOPENCL/gemm.aocx  # SGEMM computational kernel
```
[the remark this answers is not in the document]
Comment (ratusz) [reply sourceme.sh:10]: changed

File: sourceme.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+cd /home/mdokter/code/daphne-build-env/ ; source env.sh ; cd -
+
+#initialize Intel FPGA OpenCL environment
+source /opt/intel/intelFPGA_pro/21.4/hld/init_opencl.sh
```
[the remark this answers is not in the document]
Comment (ratusz) [reply sourceme.sh:5]: changed file name and hard coded paths

File: src/runtime/local/kernels/FPGAOPENCL/MatMul.h [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+#define ACL_ALIGNMENT 64
+//void *acl_aligned_malloc(size_t size) {
+//    void *result = NULL;
+//    posix_memalign(&result, ACL_ALIGNMENT, size);
+//    return result;
+//}
+    const int OUTERMOST_I = ceil(nr1/448.0);
+    const int OUTERMOST_J = ceil(nc2/512.0);
+    const int OUTERMOST_K = ceil(nc1/512.0);
```
[the remark this answers is not in the document]
Comment (ratusz) [reply src/runtime/local/kernels/FPGAOPENCL/MatMul.h:102]: changed

File: src/runtime/local/kernels/genKernelInst.py [hunk 8]
```diff
@@ -204,7 +204,7 @@ def printHelp():
         print()
         printHelp()
         sys.exit(1)
-
+    print("python test!")
```
[the remark this answers is not in the document]
Comment (ratusz) [reply src/runtime/local/kernels/genKernelInst.py:207]: removed

File: src/runtime/local/kernels/FPGAOPENCL/MatMul.h [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+    float *A, *B, *C;
+    void *aa,*bb,*cc;
+    const int TOTAL_I = III * II * OUTERMOST_I;
+    const int TOTAL_J = JJJ * JJ * OUTERMOST_J;
+    const int TOTAL_K = KKK * KK * OUTERMOST_K;
+    
+    long int num_elem_A = (long int)TOTAL_I*TOTAL_K;
+    long int num_elem_B = (long int)TOTAL_K*TOTAL_J;
+    long int num_elem_C = (long int)TOTAL_I*TOTAL_J;
```
[the remark this answers is not in the document]
Comment (ratusz) [reply src/runtime/local/kernels/FPGAOPENCL/MatMul.h:112]: ok

File: src/runtime/local/kernels/FPGAOPENCL/gemm_interface.cpp (hunk 1 above)
Comment (ratusz) [reply src/runtime/local/kernels/FPGAOPENCL/gemm_interface.cpp:132]: any additional details where exactly such function should be called?

**Review by corepointer — APPROVED** [review]

**Comment by corepointer** [comment]
LGTM - thank you for your efforts @ratusz. I merged this with slight modifications. 
* I renamed gemm.aocx to sgemm.aocx and changed the file location to a subdirectory (bitstreams)
* The sgemm bitstream is downloaded via the build script from a separate repository
* I removed the custom setup step that would source the env var script in our development VM
* I did minor code cleanup (removed some commented out code)

**Pull request merged**